OpenRA.Mods.Common/Widgets/Logic/Editor/ActorSelectorLogic.cs

```diff
@@ -0,0 +1,119 @@
+using OpenRA.Mods.Common.Graphics;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Mods.Common.Widgets;
+using OpenRA.Primitives;
+using OpenRA.Traits;
+using OpenRA.Widgets;
+
+namespace OpenRA.Mods.RA.Widgets.Logic
+{
+	public class ActorSelectorLogic
+	{
+		readonly MapEditorWidget editor;
+		readonly ScrollPanelWidget panel;
+		readonly ScrollItemWidget itemTemplate;
+		readonly Ruleset modRules;
+
+		[ObjectCreator.UseCtor]
+		public ActorSelectorLogic(Widget widget, World world, WorldRenderer worldRenderer, Ruleset modRules)
+		{
+			this.modRules = modRules;
+
+			editor = widget.Parent.Get<MapEditorWidget>("MAP_EDITOR");
+			panel = widget.Get<ScrollPanelWidget>("ACTORTEMPLATE_LIST");
+			itemTemplate = panel.Get<ScrollItemWidget>("ACTORPREVIEW_TEMPLATE");
+			panel.Layout = new GridLayout(panel);
+
+			var ownersDropDown = widget.Get<DropDownButtonWidget>("OWNERS_DROPDOWN");
+			var ownernames = world.Map.Players.Values.Select(p => p.Name);
+			Func<string, ScrollItemWidget, ScrollItemWidget> setupItem = (option, template) =>
+			{
+				var item = ScrollItemWidget.Setup(template,
+					() => ownersDropDown.Text == option,
+					() => { ownersDropDown.Text = option; editor.SelectedOwner = option; });
+				item.Get<LabelWidget>("LABEL").GetText = () => option;
+				return item;
+			};
+			ownersDropDown.OnClick = () =>
+				ownersDropDown.ShowDropDown("LABEL_DROPDOWN_TEMPLATE", 270, ownernames, setupItem);
+			ownersDropDown.Text = ownernames.First();
+			editor.SelectedOwner = ownersDropDown.Text;
+
+			IntializeActorPreview(widget, world, worldRenderer);
+		}
+
+		void IntializeActorPreview(Widget widget, World world, WorldRenderer worldRenderer)
+		{
+			panel.RemoveChildren();
+
+			var actors = modRules.Actors.Where(a => !a.Value.Name.Contains('^') && !a.Value.Name.Contains('.'))
+				.Select(a => a.Value);
+
+			foreach (var a in actors)
+			{
+				var actor = a;
+				var filter = actor.Traits.GetOrDefault<EditorTilesetFilterInfo>();
+				if (filter != null)
+				{
+					if (filter.ExcludeTilesets != null && filter.ExcludeTilesets.Contains(world.TileSet.Id))
+						continue;
+					if (filter.RequireTilesets != null && !filter.RequireTilesets.Contains(world.TileSet.Id))
+						continue;
+				}
+
+				var td = new TypeDictionary();
+				td.Add(new FacingInit(92)); // TODO: this is not forwarded when placed on the map
+				td.Add(new TurretFacingInit(92));
+
+				var player = world.Players.FirstOrDefault(p => p.InternalName == editor.SelectedOwner) ?? world.Players.First();
+
+				try
+				{
+					var item = ScrollItemWidget.Setup(itemTemplate,
+						() => actor.Name == editor.SelectedActor,
+						() => { editor.SelectedActor = actor.Name; });
+
+					var preview = item.Get<ActorPreviewWidget>("ACTOR_PREVIEW");
+					preview.SetPreview(actor, player, td);
+
+					// Scale templates to fit within the panel
+					var scale = 1f;
+					while (scale * preview.IdealPreviewSize.X > itemTemplate.Bounds.Width)
+						scale -= 0.1f;
```

Comment:
Doing this with a loop is silly then.  You can calculate the desired scale directly.